core/src/main/scala/cats/Eval.scala

```diff
@@ -134,23 +136,57 @@ final case class Now[A](value: A) extends Eval[A] {
  * by the closure) will not be retained, and will be available for
  * garbage collection.
  */
-final class Later[A](f: () => A) extends Eval[A] {
-  private[this] var thunk: () => A = f
-
-  // The idea here is that `f` may have captured very large
-  // structures, but produce a very small result. In this case, once
-  // we've calculated a value, we would prefer to be able to free
-  // everything else.
-  //
-  // (For situations where `f` is small, but the output will be very
-  // expensive to store, consider using `Always`.)
-  lazy val value: A = {
-    val result = thunk()
-    thunk = null // scalastyle:off
-    result
-  }
+final class Later[A](f: () => A) extends Eval[A] with Externalizable {
+
+  // we use lock to prevent possible problems with someone else
+  // synchronizing on 'this'.
+  private[this] val lock = new Object()
+  @volatile private[this] var thunk: Function0[A] = f
+  private[this] var result: A = _
+
+  final def value: A =
+    if (thunk == null) {
+      result
```

Comment:
I agree one should not return null, but if they do, this code will first
return null, but on a subsequent call would NPE (you would try to call
thunk again, but it would be null).

To me, this behavior sounds no good. I'd rather NPE the first time around
hen returning null once then failing if you happen to get the value again.

On Wednesday, May 4, 2016, Mike Curry <notifications@github.com> wrote:

> In core/src/main/scala/cats/Eval.scala
> <https://github.com/typelevel/cats/pull/989#discussion_r62152850>:
>
> > -  lazy val value: A = {
> > -    val result = thunk()
> > -    thunk = null // scalastyle:off
> > -    result
> > -  }
> > +final class Later[A](f: () => A) extends Eval[A] with Externalizable {
> > +
> > +  // we use lock to prevent possible problems with someone else
> > +  // synchronizing on 'this'.
> > +  private[this] val lock = new Object()
> > +  @volatile private[this] var thunk: Function0[A] = f
> > +  private[this] var result: A = _
> > +
> > +  final def value: A =
> > +    if (thunk == null) {
> > +      result
>
> I'm not sure I understand.thunk() == null should never be true, as it is
> the evaluation of a Function0[A]- which evaluates to an A. Well - at
> least no more than any other scala function with a return type of A can
> evaluate to null. So, I think thunk() == null should not be considered
> here.
>
> The reference to thunk can equal null, i.e. thunk == null, but there is
> only one dereference of thunk to and this is in the synchronized block,
> which is only entered when result == null.
>
> So, I think that the volatile nature of result would make that safe.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/typelevel/cats/pull/989/files/39828830bb958ff9bc775ed34897e72d3f59a7c6#r62152850>
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin